alfresco-search/src/main/java/org/alfresco/solr/AlfrescoCoreAdminHandler.java

```diff
@@ -838,6 +843,153 @@ private void actionREPORT(SolrQueryResponse rsp, SolrParams params, String cname
+        for(Tracker tracker : trackers)
+        {
+            if(tracker instanceof MetadataTracker)
+            {
+                metadataTracker = (MetadataTracker)tracker;
+            }
+        }
+
+        DocRouter docRouter = metadataTracker.getDocRouter();
+
+        if(docRouter instanceof DBIDRangeRouter) {
+
+            DBIDRangeRouter dbidRangeRouter = (DBIDRangeRouter) docRouter;
+            long startRange = dbidRangeRouter.getStartRange();
+            long endRange = dbidRangeRouter.getEndRange();
+
+            long maxNodeId = informationServer.maxNodeId();
+            long nodeCount = informationServer.nodeCount();
+
+            long bestGuess = -1;  // -1 means expansion cannot be done. Either because expansion
+                                 // has already happened or we're above safe range
+
+            long range = endRange - startRange; // We want this many nodes on the server
+            //System.out.println("##### Range:" + range);
+
+            long midpoint = startRange + ((long) (range * .5));
+            //System.out.println("##### Midpoint:"+midpoint);
+
+            long safe = startRange + ((long) (range * .75));
+            //System.out.println("##### Safe:"+safe);
+
+
+            long offset = maxNodeId-startRange;
+            //System.out.println("#####offset:"+offset);
+
+            //System.out.println("#####max nodeid:"+maxNodeId);
+
+            double density = 0;
+
+            if(offset > 0) {
+                density = ((double)nodeCount) / ((double)offset); // This is how dense we are so far.
+            }
+
+            //System.out.println("#####density:"+density);
+
+            if (!dbidRangeRouter.getExpanded())
+            {
+                if(maxNodeId <= safe)
+                {
+                    if (maxNodeId >= midpoint)
+                    {
+                        if(density >= 1)
+                        {
+                            //This is fully dense shard. I'm not sure if it's possible to have more nodes on the shards
+                            //then the offset, but if it does happen don't expand.
+                            bestGuess=0;
+                        }
+                        else
+                        {
+                            double multiplier = range / (density*range); // Extrapolate across the entire range to find the multiplier
+                            bestGuess = (long)(range*multiplier)-range; // This is how much to add
+                        }
+                    }
+                    else
+                    {
+                        bestGuess = 0; // We're below the midpoint so it's to early to make a guess.
+                    }
+                }
+            }
+
+            rsp.add("start", startRange);
+            rsp.add("end", endRange);
+            rsp.add("nodeCount", nodeCount);
+            rsp.add("maxDbid", maxNodeId);
+            rsp.add("density", Math.abs(density));

```

Comment:
Density must be +ve unless I have missed something.

The calculation is correct for the value to add to the current max value to get the desired number of nodes in the shard.

Reply:
Not sure I understand this comment. I'll review the code again.